ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,206 @@
+      synchronized (CompactionHeartbeatService.class) {
+        if (instance == null) {
+          LOG.debug("Initializing compaction txn heartbeater service.");
+          instance = new CompactionHeartbeatService(conf);
+          ShutdownHookManager.addShutdownHook(() -> {
+            try {
+              instance.shutdown();
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+          }, SHUTDOWN_HOOK_PRIORITY);
+        }
+      }
+    }
+    if (instance.shuttingDown) {
+      throw new IllegalStateException("The CompactionHeartbeatService is already destroyed!");
+    }
+    return instance;
+  }
+
+  private final ObjectPool<IMetaStoreClient> clientPool;
+  private volatile boolean shuttingDown = false;
+  private final long initialDelay;
+  private final long period;
+  private final ConcurrentHashMap<Long, CompactionHeartbeater> tasks = new ConcurrentHashMap<>(30);
+
+  /**
+   * Starts the heartbeat for the given transaction
+   * @param txnId The id of the compaction txn
+   * @param lockId The id of the lock associated with the txn
+   * @param tableName Required for logging only
+   * @throws IllegalStateException Thrown when the heartbeat for the given txn has already been started.
+   */
+  void startHeartbeat(long txnId, long lockId, String tableName) {
+    if (shuttingDown) {
+      throw new IllegalStateException("Service is shutting down, starting new heartbeats are not possible!");
+    }
+    if (tasks.containsKey(txnId)) {
+      throw new IllegalStateException("Heartbeat was already started for TXN " + txnId);
+    }
+    LOG.info("Submitting heartbeat task for TXN {}", txnId);
+    CompactionHeartbeater heartbeater = new CompactionHeartbeater(txnId, lockId, tableName);
+    heartbeater.start();
+    tasks.put(txnId, heartbeater);
+  }
+
+  /**
+   * Stops the heartbeat for the given transaction
+   * @param txnId The id of the compaction txn
+   * @throws IllegalStateException Thrown when there is no {@link CompactionHeartbeater} task associated with the
+   * given txnId.
+   */
+  void stopHeartbeat(long txnId) throws InterruptedException {
+    LOG.info("Stopping heartbeat task for TXN {}", txnId);
+    CompactionHeartbeater heartbeater = tasks.get(txnId);
+    if (heartbeater == null) {
+      throw new IllegalStateException("No registered heartbeat found for TXN " + txnId);
+    }
+    try {
+      heartbeater.stop();
+    } finally {
+      tasks.remove(txnId);
+    }
+  }
+
+  /**
+   * Shuts down the service, by closing its underlying resources. Be aware that after shutdown this service is no
+   * longer usable, there is no way to re-initialize it.
+   * @throws InterruptedException
+   */
+  void shutdown() throws InterruptedException {
+    shuttingDown = true;
+    LOG.info("Shutting down compaction txn heartbeater service.");
+    for (CompactionHeartbeater heartbeater : tasks.values()) {
+      heartbeater.stop();
```

Comment:
what if one of the task.stop() throws InterruptedException what happens with the rest?